nova/virt/vmwareapi/vmops.py

```diff
@@ -2253,46 +2406,161 @@ def _relocate_vm(self, vm_ref, context, instance, network_info,
+        # Due to migration.dest_host length limit, the dest doesn't contain
+        # all the needed parameters for the VMWare API connection. Thus, some
+        # parameters are picked from the local config, so it is expected that
+        # the dest compute uses the same (at least for the cacert).
+        vmware_conf.update({
+            'retry_count': CONF.vmware.api_retry_count,
+            'scheme': 'https',
+            'cacert': CONF.vmware.ca_file,
+            'insecure': CONF.vmware.insecure,
+            'pool_size': CONF.vmware.connection_pool_size
+        })
+        return self._session_cls(**vmware_conf)
+
+    def _do_migrate_disk(self, context, vm_ref, instance, dest):
+        """Copies the VM to the destination by doing a clone.
+
+        If the destination is another vCenter service, it creates a new
+        VMwareAPISession to the destination vCenter service to gather the
+        needed info for cloning the vm: datastore, service locator, folder.
+        """
+        image_meta = instance.image_meta
+        storage_policy = self._get_storage_policy(instance.flavor)
+        allowed_ds_types = ds_util.get_allowed_datastore_types(
+            image_meta.properties.hw_disk_type)
+
+        # Determine whether we should connect to the dest vCenter endpoint
+        # (if the instance is to be migrated outside of current vCenter)
+        remote = self._is_remote(dest)
+
+        service_spec = None
+        if remote:
+            session = self._create_dest_session(dest)
+            # set the instance_uuid needed in _get_service_locator_spec
+            dest.update({
+                'instance_uuid':
+                    session.vim.service_content.about.instance_uuid})
+            service_spec = self._get_service_locator_spec(dest)
+            cluster = vutil.get_moref(dest['cluster'],
+                                      'ClusterComputeResource')
+            datastore_regex = dest['datastore_regex']
+            # For migration to other vCenter service the disk_move_type
+            # needs to be moveAllDiskBackingsAndDisallowSharing
+            disk_move_type = "moveAllDiskBackingsAndDisallowSharing"
+        else:
+            session = self._session
+            cluster = self._cluster
+            datastore_regex = self._datastore_regex
+            disk_move_type = "moveAllDiskBackingsAndAllowSharing"
+
+        res_pool = vm_util.get_res_pool_ref(session, cluster)
+        datastore = ds_util.get_datastore(session, cluster,
+                                          datastore_regex,
+                                          storage_policy,
+                                          allowed_ds_types)
+        dc_info = self.get_datacenter_ref_and_name(datastore.ref)
+        folder = self._get_project_folder(dc_info, instance.project_id,
+                                          'Instances', session=session)
+
+        rel_spec = vm_util.relocate_vm_spec(self._session.vim.client.factory,
```

Comment:
`RelocateVM_Task` must always be invoked from the source vCenter.

Reply:
`RelocateVM_Task` will be invoked using `self._session` (since it has to be invoked from the source vCenter) so I think the request spec should be constructed using the same session.